### PR TITLE
DataLinks: Enable access to labels & field names

### PR DESCRIPTION
**Pull request by torkelo**
Closes https://github.com/grafana/grafana/pull/18918
Today we have:

* __series_name
* __field_name
* __value_time
* __calc (if value is reducerd to single calculation like min/max)
* __cell_xxx to access value in a different column on same row.

Some options

Option A (super duper simple):
just add 1 new variable
* __series_labels

Usage
* `${__series_labels.datacenter}`

Option B:
* __series (object)
  * name
  * labels
  * datacenter
     * host
* __field (object
  * name
  * unit

So you could write
 *  `${__series.name} ${__series.labels.datacenter}`
 *  `${__field.name} ${__field.unit}`

Option C:
single "value" object

*  `${value.seriesName} ${value.fieldName} ${value.labels.datacenter} ${value.time}`

Option D: More nested:

* `${value.series.name} ${value.field.name} ${value.series.label.datacenter} ${value.time}`

**TODO**
* [ ] figure out how to handle __cell_xxx better
* [x] How to be backward compatible with old names (done with DashboardMigrator (dd0271fa5f0402c51c4bba489c14929d9703bbff, 81c02600e5546e8544aa65f4a2ac8051393436ae)
* [ ] Retrieve suggestions lazily - will require DataLinksInput refactor, todo in separate PR
* [x] Sync with #18899 - `panelQueryRunner.getCurrentData` is removed so we need to make visualisation tab subscribe to query results as QueriesTab does (#19039)
* [x] Add test dashboard

### CODE REVIEW

**Comment by dprokop** [comment]
Data links exist in the context of a single value, that's why my thinking is that Option C & D adds unnecessary complexity to the variable syntax. Complexity that we should rather explain when describing data links.

It also implies another complexity to handle which is how are those variables going to be accessible and discoverable from the the variables suggestions popover. My first thought here is to use cascader UI https://github.com/react-component/cascader, as there might be a lot of labels available, and displaying all of the possible values as a plain list could be too much. Or maybe we could show all of the available variable choices and provide a filtering on the variables suggestions list. 

Using cascader seems like a good option to me, but then it's a question about how many levels of the cascader we wanna show, what equals to how we communicate to the user what variables can he or she use. Right now we show the actual variable names (i.e. __seriesName) in the UI with a description what this variable is going to be interpolated to. But I think we could be less detailed here and maybe only show the actual variable name **in the link** when the user has chosen the variable. So, current dropdown:
```
+---------------------------+
|__seriesName               |
|---------------------------|
|__valueTime                |
|---------------------------|
|__allVariables             |
|---------------------------|
|__otherVariab              |
+---------------------------+
```

Could become more human-readable:
```
+---------------------------+
|Series name                |
|---------------------------|
|Value time                 |
|---------------------------|
|All variables              |
|---------------------------|
|Other variable             |
+---------------------------+
```                       
                        
Then we could introduce `Labels`, which would use the cascader:

```
                                                                                       
+---------------------------+---------------------------+---------------------------+  
|Series name                |datacenter                >|host                       |  
|---------------------------|---------------------------|---------------------------|  
|Field name                 |label 1                    |                           |  
|---------------------------|---------------------------|                           |  
|Labels                    >|label 2                    |                           |  
|---------------------------|---------------------------|                           |  
|Value time                 |                           |                           |  
|---------------------------|                           |                           |  
|All variables              |                           |                           |  
|---------------------------|                           |                           |  
|Other variable             |                           |                           |  
+---------------------------+---------------------------+---------------------------+  
```
                                                                                       
And the same pattern we could use for Template variables, which would be available via the cascader.

**Comment by dprokop** [comment]
Another UI complexity that comes to my mind. What should happen when user starts editing the variable in the link? I.e. what happens when user removes two last characters in the variable and starts typing? Should we display the entire suggestions list, or only those values that are valid in this context? i.e when editing label path we probably should only show the labels that makes sense in current path's context.

**Comment by ryantxu** [comment]
> Data links exist in the context of a single value

Not true.  It could also exist in the context of a series (DataFrame) w/o the actual point.  Think about clicking on a series in the legend or similar.

Dominik -- I think a UI to edit/manage what is available may be worthwhile, but not a required first step.

What about:
```
__series.  (essentially the raw DataFrame -- maybe with all `fields` removed)
  .name
  .labels
  .refId
  .meta etc
```

When a point/reducer is picked:
```
__selected   (any better ideas?)
  .field   (the field object, maybe remove values?)
  .value  (the processed value -- numeric or text from displayValue?)
  // if it came from an actual point, not a reducer
  .row   (DataFrameView if the value is not a reducer)
  .time  //  The corresponding time 
```

When a reducer is used to show the value:
```
__calc 
```

In this case `__cell_xxx` could be automatically translated to: `__selected.row[xxx]`

**Review by ryantxu — COMMENTED** [review]
Maybe `__selected` is not the right word.

Could be:
```
const linkvar = {
  field, // All the field properties (name, config etc... remove values?)
  value: <the display number or text>
  time?: number // time corresponding to the selected value
  calc?: string // name of the reducer
  row?: any // 
}
```

just not sure what name we should expose this variable under

**Comment by torkelo** [comment]
So after discussing this 

We decided to move forward with the 3 top level variables:

* __series (contains name & labels)
* __fields (contains name & unit)
* __value (contains time & value, not sure if value is numeric or not)

**Comment by ryantxu** [comment]
> __fields (contains name & unit)

I assume you mean `__field` (singular).  `name` for sure

Do we *need* unit right now?  I ask because it would be great if we had units like [areaF2](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/utils/valueFormats/categories.ts#L65) avaliable as "ft²" and "Square Feet (ft²)", not just "areaF2"

**Comment by ryantxu** [comment]
`__value` is a little difficult -- what are its properties?  `__value.value`?

maybe it can be the row from DataFrameView?  so if the property `.time` exists, then it is available

**Comment by dprokop** [comment]
> I assume you mean __field (singular)

yes!

>__value is a little difficult -- what are its properties? __value.value?

What about `__value.[numeric|text|raw|time]` ?

File: packages/grafana-data/src/utils/dataFrameHelper.ts [hunk 1]
```diff
@@ -5,16 +5,23 @@ import { ArrayVector, MutableVector, vectorToArray, CircularVector } from './vec
 import isArray from 'lodash/isArray';
 import isString from 'lodash/isString';
 
+interface FieldCacheItem {
+  field: Field;
+  idx: number;
+}
 export class FieldCache {
-  fields: Field[] = [];
+  fields: FieldCacheItem[] = [];
```
Comment (ryantxu) [inline packages/grafana-data/src/utils/dataFrameHelper.ts]: rather than having `FieldCacheItem` with two properties... what about extending Field:
```
interface FieldCacheItem extends Field {
  index: number;
}
```

then it can just behave normally without needing the extra property reference

File: public/app/plugins/datasource/prometheus/result_transformer.ts [hunk 2]
```diff
@@ -76,7 +76,7 @@ export class ResultTransformer {
       datapoints: dps,
       query: options.query,
       target: metricLabel,
-      labels: metricData.metric,
+      tags: metricData.metric,
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/result_transformer.ts]: Is this related?  seems like something else

Comment (dprokop) [reply public/app/plugins/datasource/prometheus/result_transformer.ts]: Uhm, not sure, @torkel?

File: packages/grafana-data/src/utils/dataFrameHelper.ts (hunk 1 above)
Comment (dprokop) [reply packages/grafana-data/src/utils/dataFrameHelper.ts]: Yeah, sounds good

File: public/app/plugins/datasource/prometheus/result_transformer.ts (hunk 2 above)
Comment (torkelo) [reply public/app/plugins/datasource/prometheus/result_transformer.ts]: Tags is the old property name in legacy format , gets translated to new DataFrame.labels

**Comment by torkelo** [comment]
Looks like this could need a rebase on master, so the previous PR is not included in diff

File: public/app/plugins/datasource/prometheus/result_transformer.ts (hunk 2 above)
Comment (ryantxu) [reply public/app/plugins/datasource/prometheus/result_transformer.ts]: See https://github.com/grafana/grafana/pull/18977  I moved these changes to their own PR

**Comment by ryantxu** [comment]
> What about __value.[numeric|text|raw|time] ?

Sounds good.   I am a bit conflicted on "time" though since I do not want us to always assume X is time.

What about another variable `__row` that is the DataFrameView for that row?  Then *if* x is time it should have a value sitting at `__row.time` but this would work just as well if it is vs something other than time

File: packages/grafana-data/src/utils/dataFrameHelper.ts [hunk 3]
```diff
@@ -5,16 +5,22 @@ import { ArrayVector, MutableVector, vectorToArray, CircularVector } from './vec
 import isArray from 'lodash/isArray';
 import isString from 'lodash/isString';
 
+interface FieldWithIndex extends Field {
+  idx: number;
```
Comment (ryantxu) [inline packages/grafana-data/src/utils/dataFrameHelper.ts]: nit - consider "index" rather than "idx"

File: public/app/plugins/panel/graph/module.ts [hunk 4]
```diff
@@ -343,6 +343,10 @@ class GraphCtrl extends MetricsPanelCtrl {
   formatDate = (date: DateTimeInput, format?: string) => {
     return this.dashboard.formatDate.apply(this.dashboard, [date, format]);
   };
+
+  getDataFrameByName = (seriesName: string) => {
+    return this.dataList.filter(dataFrame => dataFrame.name === seriesName)[0];
+  };
```
Comment (ryantxu) [inline public/app/plugins/panel/graph/module.ts]: I don't totally understand how and where this is called -- but we should be careful not to assume the DataFrame name is unique.  name is an optional string.  refId may be better, but there will often be multiple DataFrames with the same refId.   maybe consider index?

Comment (dprokop) [reply public/app/plugins/panel/graph/module.ts]: This is used in graph vis to get the data frame corresponding to the series being clicked. Wondering, are multiple data frames with same refId valid in graph context?

File: public/app/features/panel/panellinks/linkSuppliers.ts [hunk 5]
```diff
@@ -1,8 +1,31 @@
...
+  labels: Labels;
+  refId?: string;
+}
+
+interface FieldVars {
+  name: string;
+}
+
+interface ValueVars {
+  raw: number;
```
Comment (dprokop) [inline public/app/features/panel/panellinks/linkSuppliers.ts]: @ryantxu do you think we need this, or `numeric` is just fine?

Comment (ryantxu) [reply public/app/features/panel/panellinks/linkSuppliers.ts]: raw is any

File: public/app/features/panel/panellinks/linkSuppliers.ts [hunk 6]
```diff
@@ -1,8 +1,31 @@
 import { PanelModel } from 'app/features/dashboard/state/PanelModel';
-import { FieldDisplay, ScopedVars, DataLinkBuiltInVars } from '@grafana/ui';
-import { LinkModelSupplier, getTimeField } from '@grafana/data';
+import { FieldDisplay, ScopedVars, ScopedVar } from '@grafana/ui';
+import { LinkModelSupplier, getTimeField, Labels } from '@grafana/data';
 import { getLinkSrv } from './link_srv';
 
+interface SeriesVars {
+  name: string;
```
Comment (ryantxu) [inline public/app/features/panel/panellinks/linkSuppliers.ts]: DataFrame name is optional... or will this fill things in with a default?
same with labels...

File: public/app/features/panel/panellinks/linkSuppliers.ts [hunk 7]
```diff
@@ -14,27 +37,44 @@ export const getFieldLinksSupplier = (value: FieldDisplay): LinkModelSupplier<Fi
...
           if (timeField) {
-            scopedVars[DataLinkBuiltInVars.valueTime] = {
-              text: 'Value time',
-              value: timeField.values.get(value.row),
+            scopedVars['__value'] = {
+              value: {
+                raw: field.values.get(value.rowIndex),
+                numeric: value.display.numeric,
+                text: value.display.text,
+                time: timeField.values.get(value.rowIndex),
```
Comment (ryantxu) [inline public/app/features/panel/panellinks/linkSuppliers.ts]: This should not be nested under `if (timeField) {`

everything should exist except the time property when there is no time field

File: public/app/plugins/panel/graph/module.ts (hunk 4 above)
Comment (ryantxu) [reply public/app/plugins/panel/graph/module.ts]: > Wondering, are multiple data frames with same refId valid in graph context?

For sure, in influx the query `SELECT a,b FROM measurement` is returned as two TimeSeries with the same refId.

File: public/app/features/panel/panellinks/linkSuppliers.ts [hunk 8]
```diff
@@ -1,8 +1,31 @@
...
+  raw: number;
+  numeric: number;
+  text: string;
+  time?: number;
+}
+
+interface DataLinkScopedVars extends ScopedVars {
+  __series?: ScopedVar<SeriesVars>;
+  __field?: ScopedVar<FieldVars>;
+  __value?: ScopedVar<ValueVars>;
```
Comment (ryantxu) [inline public/app/features/panel/panellinks/linkSuppliers.ts]: can we add `__row?:ScopedVar<any>;`

The row would be backed by DataFrameView and can be accessed either by name or index:
`__row[0]` or `__row.time`

then we could also replace any `__cell_1` variable with `__row[1]`

Comment (dprokop) [reply public/app/features/panel/panellinks/linkSuppliers.ts]: Yes!

File: public/app/plugins/panel/graph/module.ts (hunk 4 above)
Comment (dprokop) [reply public/app/plugins/panel/graph/module.ts]: Thanks for clarification

File: packages/grafana-ui/src/utils/dataLinks.ts [hunk 9]
```diff
@@ -4,8 +4,11 @@ import { LinkModelSupplier } from '@grafana/data';
 export const DataLinkBuiltInVars = {
   keepTime: '__url_time_range',
   includeVars: '__all_variables',
-  seriesName: '__series_name',
-  valueTime: '__value_time',
+  seriesName: '__series.name',
+  fieldName: '__field.name',
+  valueTime: '__value.time',
+  valueNumeric: '__value.numeric',
+  seriesRefId: '__series.refId',
```
Comment (torkelo) [inline packages/grafana-ui/src/utils/dataLinks.ts]: Can't think of a use case fo refId just yet, so I think we can wait until we do.

**Review by torkelo — COMMENTED** [review]
Looks really good. For the auto-complete editor tooltip, I think it's a bit too segmented (sub lists with headings) like two entries with Name (one under series and one under Field). think we need to combine these a bit more to "Series name" "Field name" to make the list a little easier to scan. But we can talk about that on meet.

**Comment by dprokop** [comment]
>  But we can talk about that on meet.

I'm not super happy with that neither. Working on improvements to that now

File: public/app/core/time_series2.ts [hunk 10]
```diff
@@ -75,7 +75,12 @@ export function getDataMinMax(data: TimeSeries[]) {
 export default class TimeSeries {
   datapoints: any;
   id: string;
+  // Represents index of original data frame in the quey response
   dataFrameIndex: number;
+  /**
+   * Name of the field the time series was created from
+   * Used in graph panel to retrieve value from the original data frame
+   */
   fieldName: string;
```
Comment (ryantxu) [inline public/app/core/time_series2.ts]: As I think more... should we really store the fieldIndex?  name does not need to be unique (though it most likely is) and requires scanning all fields to find it

Comment (dprokop) [reply public/app/core/time_series2.ts]: as we discussed -> to be replaced with fieldIndex

File: packages/grafana-ui/src/components/DataLinks/DataLinksEditor.tsx [hunk 11]
```diff
@@ -47,6 +47,9 @@ export const DataLinksEditor: FC<DataLinksEditorProps> = React.memo(({ value, on
 
   return (
     <>
+      {/* <div className="grafana-info-box"> */}
```
Comment (dprokop) [inline packages/grafana-ui/src/components/DataLinks/DataLinksEditor.tsx]: @torkelo to be removed?

**Review by dprokop — APPROVED** [review]
